### ISSUE TITLE
Investigate Webpack

### ISSUE THREAD

**Issue by chuckharmston** [mozilla/testpilot#1463]
Our gulp build process is a bit gnarly and DIY, and some things (i.e. #1376, #1349) are much more challenging to do than it would be when using Webpack. Let's put together a list of requirements for our build system in a post-Django world, and do some informed compare/contrasting.
## Requirements
- Should be entirely contained within the `frontend` directory, as `addon` is.
- Development server
  - Should watch for changes and selectively recompile.
  - With hot reloading for React components.
  - With React and Redux dev tools.
- Assets
  - Should copy locale files, addon XPI and addon RDF to the appropriate locations.
  - Should convert content yaml to JSON and copy to the appropriate location.
  - Should convert legal md to HTML and copy to the appropriate location.
  - Should bundle CSS modules + SCSS.
  - Should bundle Javascript.
    - [Tree-shaking?](https://blog.engineyard.com/2016/tree-shaking)
  - Vendor modules: babel-polyfill, l20n
  - Should build source maps for JS and CSS
  - Should selectively include Pontoon.
  - Should ship with a base configuration and per-environment overrides using `NODE_ENV`
- Linting
  - SCSS and JavaScript
  - Should fail builds on linting errors, notify of warnings.
## Future
- Isomorphic; include experiments JSON blob in `index.html`, prevent need for API call.
## Proposed file structure
- addon
- frontend
  - bin
    - server.js
  - build
    - webpack.config.js
  - config
    - <env>.js
    - index.js
  - dist
    - static
      - (mirrors src)
    - app.css
    - app.js
    - index.html
    - vendor.js
  - server
  - src
    - actions
    - components
    - containers
    - lib
    - reducers
    - selectors
    - static
      - addon
        - addon.xpi
      - images
      - locales
    - app.js
    - index.html
    - routes.js
  - test
    - (mirrors src)

**Comment by chuckharmston** [comment]
Started adding notes above. Would very much appreciate double-checking by people who know better.

**Comment by chuckharmston** [comment]
Experimentalish thing going on over here: https://github.com/chuckharmston/testpilot/compare/master...chuckharmston:webpack-of-lies

**Comment by chuckharmston** [comment]
I've got a (mostly) working port of this over here, with many ideas based on the [react-redux-starter-kit](https://github.com/davezuko/react-redux-starter-kit/): https://github.com/chuckharmston/testpilot-webpack

I touched as little of the application as possible, instead focusing on the build system. Hot reloading is mostly-working, and there are some neat webpack features, such as skipping out on the experiments API call entirely, instead loading JSON directly from the YAML: https://github.com/chuckharmston/testpilot-webpack/blob/master/frontend/src/actions/experiments.js#L8-L16

**Comment by chuckharmston** [comment]
This is bitrotting pretty hard, would appreciate some eyes.

**Comment by pdehaan** [comment]
> Should be entirely contained within the frontend directory, as addon is.

Does that conflict with the goals of https://github.com/mozilla/testpilot/issues/1486?
I wasn't sure if the goal was to have a shared package.json versus needing to do a `cd addon && npm install && cd ../frontend && npm install` and then have to manage package.json version numbers in 2 places, etc.

**Comment by pdehaan** [comment]
currently if i run `$ npm run lint` from the frontend/ directory, i get about 1700 errors. Running `$ ./node_modules/.bin/eslint . --fix` auto fixes the code to match the rules (for better or worse) and reduces that number down to a reasonable-ish 182 errors. Most of which seem to be `react/prop-types` errors or `max-len` violations (or a few other random ones).

**Comment by chuckharmston** [comment]
> Does that conflict with the goals of #1486?

I do think we want two separate package.json files. They're two fully-separate codebases with different build processes. CI should be building both, but they should be separate tasks.

**Comment by chuckharmston** [comment]
@pdehaan it's a light proposal right now intended for discussion fodder, don't worry too much about the actual code :)

**Comment by pdehaan** [comment]
I also see this error when I try and run this:

```
[nodemon] starting `node bin/server index.js`
  app:config Creating default configuration. +0ms
  app:config Looking for environment overrides for NODE_ENV "development". +3ms
  app:config Found overrides, applying to default configuration. +2ms
module.js:457
    throw err;
    ^

Error: Cannot find module '../build/webpack.config'
    at Function.Module._resolveFilename (module.js:455:15)
    at Function.Module._load (module.js:403:25)
    at Module.require (module.js:483:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/pdehaan/dev/tmp/del/testpilot-webpack/frontend/server/main.js:4:23)
    at Module._compile (module.js:556:32)
    at Object.Module._extensions..js (module.js:565:10)
    at Module.load (module.js:473:32)
    at tryModuleLoad (module.js:432:12)
    at Function.Module._load (module.js:424:3)
[nodemon] app crashed - waiting for file changes before starting...
```

Looks like we may be missing the ./frontend/build/ directory in the repo.

**Comment by lmorchard** [comment]
Haven't checked out the code from your branch yet, but some general additions / comments...
- Produce `index.html` loading stubs for all pages (especially experiments) that include social sharing metadata in `<head>`.
- Local dev server should use HTTPS
- Probably just an expansion on "Isomorphic", but I'd require: Can render React-based pages to static index.html as a head-start in rendering (#1306). I've queued up enough unread dev blog posts in Pocket to guess that this is totally doable in Webpack.
- We'll need to keep the `experiments.json` resource around for the foreseeable future, since we do actually have other teams at Mozilla who use it. In particular, the measurement team. That could change / we could negotiate a different mechanism.

> skipping out on the experiments API call entirely, instead loading JSON directly from the YAML

This is neat, though I wonder if we get similar benefits out of keeping the JSON and adding static rendering (#1306)? 

But I'm kind of splitting hairs, because it's not like we have so much YAML content that it would hurt to do all of the above via automated build. (i.e. render static pages _and_ roll the YAML content into the `app.js` _and_ keep producing the JSON as long as we need it)

> I do think we want two separate package.json files. They're two fully-separate codebases with different build processes. CI should be building both, but they should be separate tasks.

I do think we want separate `package.json` files. But, the idea in #1486 is that it would be nice to have a single-command way to run both of those build & file watching tasks with a single command for better dev convenience. I think that can be done with separate `package.json` files, though.

**Comment by lmorchard** [comment]
Thinking this is just another :+1: for tree-shaking, but I'm also curious about [rollup](http://rollupjs.org/). Though, I've read that webpack folks are looking into stealing some of their ideas.

**Comment by johngruen** [comment]
@chuckharmston based on our discussion the other day, it seems like the investigation has ended and we're gonna stick with gulp for awhile. Can we go ahead and close this?

**Comment by chuckharmston** [comment]
Yeah, closing this. I investigated a bunch of webpack approaches, and the benefits just don't outweigh the costs (which would be substantial) to make this switch. It's worth considering in any future rewrites.
